packages/patternfly-4/react-core/src/components/Nav/Nav.js

```diff
@@ -0,0 +1,119 @@
+import React from 'react';
+import styles from '@patternfly/patternfly-next/components/Nav/styles.css';
+import { css } from '@patternfly/react-styles';
+import PropTypes from 'prop-types';
+import { NavVariants } from './NavVariants';
+
+const propTypes = {
+  /** Indicates the type of list. Only used for flat lists. */
+  variant: PropTypes.oneOf(Object.values(NavVariants)),
+  /** NavList elements */
+  children(props, propName, componentName) {
+    const prop = props[propName];
+
+    let error = null;
+    React.Children.forEach(prop, child => {
+      if (child.type.componentType !== 'NavList') {
+        error = new Error(`${componentName} children should be of type NavList.`);
+      }
+    });
+    return error;
+  },
+  /** Additional classes added to the Nav */
+  className: PropTypes.string,
+  /** Callback for updating when selection changes */
+  onSelect: PropTypes.func,
+  /** Callback for when a list is expanded or collapsed */
+  onToggle: PropTypes.func
+};
+
+const defaultProps = {
+  variant: 'default',
+  children: null,
+  className: '',
+  onSelect: () => undefined,
+  onToggle: () => undefined
+};
+
+const renderLists = (children, onSelect, onToggle) =>
+  React.Children.map(children, (childNode, index) =>
+    React.cloneElement(childNode, {
+      groupId: index,
+      onSelect,
+      onToggle
+    })
+  );
+
+class Nav extends React.Component {
+  // Callback from NavListItem
+  onSelect(groupId, itemId, event) {
+    this.props.onSelect({
+      itemId,
+      groupId,
+      event
+    });
+  }
+
+  // Callback from NavGroup
+  onToggle(groupId, isExpanded, event) {
+    this.props.onToggle({
+      groupId,
+      isExpanded,
+      event
+    });
+  }
+
+  render() {
+    const { variant, children, className, ...props } = this.props;
+
+    const getVariantStyle = type =>
+      ({
+        [NavVariants.default]: styles.navList,
+        [NavVariants.simple]: styles.navSimpleList,
+        [NavVariants.horizontal]: styles.navHorizontalList,
+        [NavVariants.tertiary]: styles.navTertiaryList
+      }[type]);
+
+    // variant style is only customizable if it's a single flat list
+    let variantStyle;
+    // Figure out if we have groups
+    let isGroup = false;
+    if (Array.isArray(children)) {
+      variantStyle = styles.navList;
+      isGroup = children[0].props.title && !children[0].props.isExpandable;
+    } else {
+      variantStyle = children.props.title ? styles.navList : getVariantStyle(variant);
+      isGroup = children.props.title && !children.props.isExpandable;
+    }
+
+    let innerNav;
+    if (isGroup) {
+      innerNav = renderLists(
```

Comment:
This copying children should be avoided, I used it in dropdown as well and am planing to remove it. It's not wrong, it just makes the rendering login a bit harder and it takes more time to render. To pass down parent's onToggle function you could use [react-context](https://reactjs.org/docs/context.html) so only parent `Nav` component will have `onToggle` function and child components can trigger such function as well. Here is great example https://hackernoon.com/how-to-use-the-new-react-context-api-fce011e7d87

Reply:
I added context for the callback functions, but I still need to clone children to add indices.

Reply:
You might want to check out https://reactjs.org/docs/react-api.html#reactchildren children structure should not be assumed. You will also want to us the isValidElement check if you plan to dive into props. It is possible a user will return null, undefined, false, stings, numbers, or functions all of which will throw on the property access.